### ISSUE TITLE
Add GPU support

### ISSUE THREAD

**Issue by cliftonarms** [CodeReclaimers/neat-python#116]
It would be a real shame not to include the use of GPUs.

https://weeraman.com/put-that-gpu-to-good-use-with-python-e5a437168c01

Great code by the way.

**Comment by drallensmith** [comment]
If it could be done such that the library (as opposed to example) code
would work both with the basic python libraries only and with numpy &
related extensions when available, that could work. I have been examining
CMA-ES (purecma.py) as a possibility for this, for instance.

-Allen

**Comment by cliftonarms** [comment]
Would be awesome if you could, as your Neat implementation would "fly" with a little matrix GPU assistance. May even outperform the SharpeNeat version.

**Comment by evolvingmeat** [comment]
I was looking at OpenCL to implement with neat, but eventually give up on that idea, since most problems at least in my case is not waiting for 3+ hours, but make sense of results. Additionally it fairly cheap to get lots of CPU's to speed up simulation, as example google offers $300 of computing power on there cloud, that's a lot of computing :)
In what cases GPU speed is required for neat-python ?

**Comment by cliftonarms** [comment]
Speeding up with GPU is relatively simple if the cuda is being used. All it would need is a couple of the matrix intensive calculations being shipped to the gpu and training would be accelerated 10x. 

See here for a nice example https://weeraman.com/put-that-gpu-to-good-use-with-python-e5a437168c01

I am trying to use neat out on stock prediction, 20,000 rows of 100 inputs. Even using parallel on 8 cors it takes 24 hours to run.

**Comment by mdalvi** [comment]
As @drallensmith said, it could be done such that the code works with both, basic python libraries and with numpy and related extensions. One such extension that could serve is [cudamat](https://github.com/cudamat/cudamat), for example, see [here](http://stsievert.com/blog/2016/07/01/numpy-gpu/).

**Comment by mdalvi** [comment]
@cliftonarms since you mentioned that you are **using NEAT to beat stock market**, this research paper might be useful for you, [Designing Safe, Profitable Automated Stock Trading Agents Using Evolutionary Algorithms](http://www.cs.utexas.edu/users/ai-lab/?GECCO06-trading)

This may not be related to the issue but just curious, what kind of network you are building takes 100 inputs? I am sure you must have taken care that your information is fairly uncorrelated.

**Comment by cliftonarms** [comment]
@mdalvi I'm not building a network, I am letting NEAT do that. ;o)

I am fed up with the diminishing gradient problems of deep neural nets. I have tried evolutionary strategies which remove the problem of diminishing gradients but still require the technician to select a network topology.

This is why I am trying the NEAT method, early days yet. Good thing about NEAT is it learns to dis-regards useless inputs so its more tolerant to lots of input variables.

Thanks for the article link.

**Comment by AlconDivino** [comment]
Well i have tried something similar with Crypto trading.
i used the Numba module and made the eval_genoms() function running on gpu power and it reduces processing time significant.

```
from numba import jit
    @jit(nopython=False, parallel=True)
    def eval_genomes(self, genomes, config):
        # call eval_genome for every genome
        for genome_id, genome in genomes:
            genome.fitness = self.eval_genome(genome, config)
```
You have to use numpy tho. pandas and many other libraries won't work

**Comment by coryryan269** [comment]
Am I correct that you could import numba and decorate certain functions with @vectorize and that should improve performance considerably? I'm a little out of my depth here, but would that work?

**Comment by The-Odor** [comment]
Are there any thoughts on adding GPU support?
